### DISCUSSION TITLE
How to get Path affinity fields out of the sleap python library using a bottomuppredictor

### DISCUSSION THREAD

**Discussion by Stormlord2001** [talmolab/sleap#1576] [Help!] [answered]
It is my understanding from looking at the documentation of sleap that is should be possible to get path affinity fields out of sleap.
I have set the return pafs to true.
![image](https://github.com/talmolab/sleap/assets/117979580/5dca4d17-b5f0-47e9-9ee7-52e336780269)
and the documentation says that the path affinity fields will be returned together with the predicted instances
![image](https://github.com/talmolab/sleap/assets/117979580/fb03973c-2705-4c60-b26b-22a8eb992661)
but from my testing i am not able to se how they are returned with the predicted_instances.

I also found the following, but i am not able to find the "part_affinity_key"
![image](https://github.com/talmolab/sleap/assets/117979580/c7462cdd-12ec-4392-b4c7-1eeef7b9727b)

**Comment by roomrys** [comment]
Hi @Stormlord2001,

`Predictor.predict` is a bit too high-level and only returns the predictions. `Predictor.InferenceModel.predict` should allow returning the PAFs as well.

Here we see that the pafs are added to the `out` dictionary in the `BottomUpInferenceLayer`:
https://github.com/talmolab/sleap/blob/dbe14a8f860f2aecfdd34bce802cde03db6be000/sleap/nn/inference.py#L2931-L2932
and `InferenceModel.predict` returns the entire `outs` dictionary:
https://github.com/talmolab/sleap/blob/dbe14a8f860f2aecfdd34bce802cde03db6be000/sleap/nn/inference.py#L1025-L1033

Thanks,
Liezl

**Reply by Stormlord2001** [reply]
Hi Liezl,
Thanks for the fast response, i have tried to change the predictions into the following:'
![image](https://github.com/talmolab/sleap/assets/117979580/272fd875-0d19-44ad-b2d0-f2682c0d231a)
but the return from the function is a dict, and does not contain a "out", and the keys and from keys, does not return anything.
If you can spot what i have done wrong i would be vary greatfull.

Thanks,
Rasmus

**Reply by roomrys** [reply, marked as answer]
Hi @Stormlord2001,

>but the return from the function is a dict, and does not contain a "out", and the keys and from keys, does not return anything.

The return is the `out` dict.
```python
test["part_affinity_fields"]
```
should give you the PAFs. Check `test.keys()` for the other keys.

Thanks,
Liezl
